Improve Voip pipeline stability

jaminh opened the pull request:
It appears the pipeline is being unexpectedly cancelled in some instances. In order to mitigate this issue hang ups will be detected using a separate task rather than relying on timeouts in the STT read method. Also reading STT events will be retried once if it is cancelled. The pipeline will also catch and log any CancelledErrors to help with further debugging.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: https://github.com/home-assistant/core/issues/128372
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

jaminh committed:
Improve Voip pipeline stability

It appears the pipeline is being unexpectedly cancelled in some
instances. In order to mitigate this issue hang ups will be detected
using a separate task rather than relying on timeouts in the STT read
method. Also reading STT events will be retried once if it is cancelled.
The pipeline will also catch and log any CancelledErrors to help with
further debugging.

jaminh committed:
Update Voip tests

jaminh committed:
Remove unnecessary changes

Remove unnecessary logging and cancelled error handling in wyoming STT.

jaminh commented:
I removed check in wyoming STT for cancelled error because I think that was a red herring. I suspect just switching to the separate task to check for the phone hanging up was sufficient to stop the spurious cancelled errors while running the pipeline.

CloCkWeRX commented:
>           fd_event_list = self._selector.poll(timeout, max_ev)
E           Failed: Timeout >9.0s

On the assumption this is just CI flakey, perhaps rebase this PR to current?

jaminh commented:
All checks passing now after rebase.

jaminh committed:
Remove comment about clearing system prompt

The test no longer checks for clearing the system prompt. Since that
logic exists completely in the assist_satellite component I think it is
reasonable to only test that logic in the unit tests for that component.

jaminh commented on tests/components/voip/test_voip.py:
```diff
@@ -1120,54 +1145,21 @@ async def test_start_conversation_user_doesnt_pick_up(
-        satellite.transport = Mock()
+        satellite.connection_made(Mock())
 
-        # Error should clear system prompt
```
Since the system prompt logic is just in the assis_satellite component I thought it was reasonable to remove those checks from here. It should be sufficient to just test that the error is raised here.

bdraco commented on homeassistant/components/voip/assist_satellite.py:
```diff
@@ -234,53 +235,74 @@ async def _do_announce(
+
+                await asyncio.sleep(_HANGUP_SEC / 2)
+        except asyncio.CancelledError:
+            _LOGGER.debug("Check hangup cancelled")
```
This will swallow cancellation even if the current task is cancelling

bdraco replied:
```suggestion
            # Don't swallow cancellation
            if (current_task := asyncio.current_task()) and current_task.cancelling():
                raise
            _LOGGER.debug("Check hangup cancelled")
```

jaminh committed:
Re-raise cancellation

Re-raise CancelledError if the current task is cancelling in the check hangup task

Co-authored-by: J. Nick Koston <nick@koston.org>

jaminh commented on homeassistant/components/voip/assist_satellite.py:
```diff
@@ -354,7 +406,11 @@ async def stt_stream():
+            _LOGGER.exception("Timeout error")
             self.disconnect()  # caller hung up
+        except asyncio.CancelledError:
+            _LOGGER.debug("Pipeline cancelled")
```
Should I re-raise the CancelledError here if the current task is cancelling as well?

bdraco replied:
Yes

jaminh committed:
Re-raise CancelledError in pipeline as well

bdraco commented:
Probably needs a review from Paulus or Mike as I'm not familiar with this code enough for scope of these changes

jaminh committed:
Fix formatting issue

IceBotYT commented:
I tested this with my Poly CCX 400 + FreePBX and works great for me :)

jaminh committed:
Remove unnecessary logging

balloob committed:
Merge branch 'dev' into voip-pipeline-stability

jaminh committed:
Merge branch 'dev' into voip-pipeline-stability

jaminh committed:
Add MockResultStream import to tests

This was presumably missed while merging

jaminh committed:
Merge branch 'dev' into voip-pipeline-stability

jaminh committed:
Merge branch 'dev' into voip-pipeline-stability

balloob commented on homeassistant/components/voip/assist_satellite.py:
```diff
@@ -332,6 +357,23 @@ async def async_start_conversation(
+    def disconnect(self):
+        """Server disconnected."""
+        super().disconnect()
+        if self._check_hangup_task is not None:
```
Should we also cancel the task when it's not done?

jaminh replied:
Yes, added.

jaminh committed:
Cancel check hangup task on disconnect

jaminh committed:
Merge branch 'dev' into voip-pipeline-stability

balloob commented on homeassistant/components/voip/assist_satellite.py:
```diff
@@ -274,53 +275,77 @@ async def _do_announce(
                 self._announcement = None
-                self._check_announcement_ended_task = None
-                self._announcement_future.set_exception(
+                self._check_announcement_pickup_task = None
```
Should we also cancel this task?

balloob replied:
we always should make sure that any task we spawn, stops or is cancelled. You can check if a task is done by calling `.done()` on it.

jaminh replied:
In this case that task is what is currently executing, and will soon end at this point. It seems odd that a task would cancel itself instead of just let itself terminate normally?

jaminh committed:
Merge branch 'dev' into voip-pipeline-stability

jaminh committed:
Add myself as codeowner for VoIP

jaminh committed:
Update CODEOWNERS

jaminh commented:
@balloob I think I added myself as a code owner, should we update the voip-utils version to 0.3.2 in this PR as well or make that a separate change since these changes don't actually depend on it?

balloob commented:
We should do a standalone PR for dependency bumps.

balloob commented:
btw, 1 open comment still on this PR. https://github.com/home-assistant/core/pull/137620#discussion_r2073368334
